pkg/frontend/querymiddleware/astmapper/rangemapper.go

```diff
@@ -0,0 +1,543 @@
+// but the inner expressions can still be splittable
+var supportedVectorAggregators = map[parser.ItemType]struct{}{
+	parser.AVG:   {},
+	parser.COUNT: {},
+	parser.MAX:   {},
+	parser.MIN:   {},
+	parser.SUM:   {},
+	parser.TOPK:  {},
+}
+
+var splittableVectorAggregators = map[parser.ItemType]struct{}{
+	parser.MAX: {},
+	parser.MIN: {},
+	parser.SUM: {},
+}
+
+// Supported range vector aggregators
+
+type RangeVectorName string
+
+// TODO: are there better constant values to use here?
+const avgOverTime = RangeVectorName("avg_over_time")
+const countOverTime = RangeVectorName("count_over_time")
+const increase = RangeVectorName("increase")
+const maxOverTime = RangeVectorName("max_over_time")
+const minOverTime = RangeVectorName("min_over_time")
+const rate = RangeVectorName("rate")
+const sumOverTime = RangeVectorName("sum_over_time")
+
+var splittableRangeVectorAggregators = map[RangeVectorName]struct{}{
+	avgOverTime:   {},
+	countOverTime: {},
+	maxOverTime:   {},
+	minOverTime:   {},
+	rate:          {},
+	sumOverTime:   {},
+}
+
+// NewRangeMapper creates a new query range mapper.
+func NewRangeMapper(interval time.Duration, logger log.Logger) (ASTMapper, error) {
+	rangeMapper := NewASTNodeMapper(&rangeMapper{splitByInterval: interval})
+	return rangeMapper, nil
+}
+
+// MapNode returns node mapped as embedded queries
+func (r *rangeMapper) MapNode(node parser.Node, stats *MapperStats) (mapped parser.Node, finished bool, err error) {
+	if !isSplittable(node) {
+		// If no node in the tree is splittable, finish the AST traversal
+		return node, true, nil
+	}
+
+	// Immediately clone the node to avoid mutating the original
+	node, err = cloneNode(node)
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		return r.mapAggregatorExpr(n, stats)
+	case *parser.BinaryExpr:
+		return r.mapBinaryExpr(n, stats)
+	case *parser.Call:
+		return r.mapCall(n, stats)
+	case *parser.ParenExpr:
+		return r.mapParenExpr(n, stats)
+	// TODO: add other expression types? EvalStmt, Expressions, StepInvariantExpr, TestStmt, UnaryExpr?
+	default:
+		return n, false, nil
+	}
+}
+
+// copyWithEmbeddedExpr clones a rangeMapper with a new embedded expression.
+// This expression is the one that will be used in all the embedded queries in the split and squash operation
+func (r *rangeMapper) copyWithEmbeddedExpr(embeddedExpr *parser.AggregateExpr) *rangeMapper {
+	rangeMapper := *r
+	rangeMapper.embeddedAggregatorExpr = embeddedExpr
+	return &rangeMapper
+}
+
+// isSplittable returns whether it is possible to optimize the given
+// sample expression.
+// A vector aggregation is splittable, if the aggregation operation is
+// supported and the inner expression is also splittable.
+// A range aggregation is splittable, if the aggregation operation is
+// supported.
+// A binary expression is splittable, if at least one operand is splittable.
+func isSplittable(node parser.Node) bool {
```

Comment:
Very opinionated nitpick, but I find this comments hard to read, why do they break at ~75 chars? These could be just 4 lines reading perfectly well.

```suggestion
// isSplittable returns whether it is possible to optimize the given sample expression.
// A vector aggregation is splittable, if the aggregation operation is supported and the inner expression is also splittable.
// A range aggregation is splittable, if the aggregation operation is supported.
// A binary expression is splittable, if at least one operand is splittable.
func isSplittable(node parser.Node) bool {
```